e2etest/bgptests/cordon_node.go

```diff
@@ -0,0 +1,147 @@
+
+	"go.universe.tf/e2etest/pkg/frr"
+	frrconfig "go.universe.tf/e2etest/pkg/frr/config"
+	"go.universe.tf/e2etest/pkg/ipfamily"
+	testservice "go.universe.tf/e2etest/pkg/service"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+var _ = ginkgo.Describe("Cordon Node FRR", func() {
+	var (
+		cs            clientset.Interface
+		testNamespace string
+	)
+
+	ginkgo.BeforeEach(func() {
+		err := ConfigUpdater.Clean()
+		Expect(err).NotTo(HaveOccurred(), "cleaning k8s api CRs failed")
+
+		for _, c := range FRRContainers {
+			err := c.UpdateBGPConfigFile(frrconfig.Empty)
+			Expect(err).NotTo(HaveOccurred(),
+				fmt.Sprintf("cleaning frr config at %s failed", c.Name))
+		}
+
+		cs = k8sclient.New()
+		testNamespace, err = k8s.CreateTestNamespace(cs, "bgp")
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	ginkgo.AfterEach(func() {
+		if ginkgo.CurrentSpecReport().Failed() {
+			dumpBGPInfo(ReportPath, ginkgo.CurrentSpecReport().LeafNodeText, cs, testNamespace)
+			k8s.DumpInfo(Reporter, ginkgo.CurrentSpecReport().LeafNodeText)
+		}
+		err := k8s.DeleteNamespace(cs, testNamespace)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	ginkgo.Describe("When create an IPV4 service and then cordon node", func() {
+		var (
+			svc                *corev1.Service
+			allNodes           *v1.NodeList
+			schedulableNodeIPs []net.IP
+		)
+
+		ginkgo.BeforeEach(func() {
+			if BGPMode == "frr" {
+				for _, c := range FRRContainers {
+					c.NeighborConfig.GracefulRestart = true
+				}
+			}
+
+			_, svc = setupBGPService(cs, testNamespace, ipfamily.IPv4, []string{v4PoolAddresses}, FRRContainers, func(svc *corev1.Service) {
+				testservice.TrafficPolicyCluster(svc)
+			})
+			testservice.ValidateDesiredLB(svc)
+
+			var err error
+			allNodes, err = cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			for _, c := range FRRContainers {
+				validateService(svc, allNodes.Items, c)
+			}
+
+			err = k8s.CordonNode(cs, &allNodes.Items[0])
+			Expect(err).NotTo(HaveOccurred(), "k8s api call to cordon node failed")
+			Eventually(func() bool {
+				ret, err := k8s.IsNodeCordoned(cs, &allNodes.Items[0])
+				Expect(err).NotTo(HaveOccurred())
+				return ret
+			}, 30*time.Second, time.Second).Should(BeTrue(), "node.spec.Unschedulable not true")
+
+			schedulableNodeIPs = k8s.NodesIPv4(allNodes.Items[1:])
+		})
+
+		ginkgo.AfterEach(func() {
+
+			err := k8s.UnCordonNode(cs, &allNodes.Items[0])
+			Expect(err).NotTo(HaveOccurred())
+
+			for _, c := range FRRContainers {
+				validateService(svc, allNodes.Items, c)
+			}
+
+			testservice.Delete(cs, svc)
+		})
+
+		//			ginkgo.It("speaker on cordoned node should have empty frr config", func() {})
+		ginkgo.It("bgp peers should not include cordoned node IP as next hop", func() {
+			Eventually(func() error {
+				for _, c := range FRRContainers {
```

Comment:
I think you can use `checkServiceNotOnNodes` here to ensure that the cordoned node is not announcing.
Also, I'd try to uncordon it to see if the service is restored.

Reply:
Using that function  is a very good idea and works with VRFs which is tricky because you have to take IPs from the containers. Thanks

>  I'd try to uncordon it to see if the service is restored.
I do test it in the after each
```
		ginkgo.AfterEach(func() {

			err := k8s.UnCordonNode(cs, &allNodes.Items[0])
			Expect(err).NotTo(HaveOccurred())

			for _, c := range FRRContainers {
				validateService(svc, allNodes.Items, c)
			}

			testservice.Delete(cs, svc)
		})


```
but I would not object that this might be a test case on its own

Reply:
fair enough then